pkg/core/interop/contract/call.go

```diff
@@ -122,8 +122,44 @@ func callExFromNative(ic *interop.Context, caller util.Uint160, cs *state.Contra
 		initOff = md.Offset
 	}
 	ic.Invocations[cs.Hash]++
-	ic.VM.LoadNEFMethod(&cs.NEF, caller, cs.Hash, ic.VM.Context().GetCallFlags()&f,
-		hasReturn, methodOff, initOff, cb)
+	f = ic.VM.Context().GetCallFlags() & f
+
+	needWrap := f&(callflag.All^callflag.ReadOnly) != 0 // If the method is safe, then it's read-only and doesn't perform storage changes or emit notifications.
+	if !needWrap && ic.VM.Context().HasTryBlock() {     // If the method is not wrapped into try-catch block, then changes should be discarded anyway if exception occurs.
+		needWrap = true
+	}
+	if needWrap && ic.DAO != nil {
+		ic.DAO = ic.DAO.GetPrivate()
```

Comment:
Right here, before you wrap you know exactly the number of existing notifications (`len(ic.Notifications)`)...

Reply:
Nice solution, and it works properly for all cases, implemented.